launcher/minecraft/mod/tasks/LocalWorldSaveParseTask.cpp

```diff
@@ -0,0 +1,180 @@
+
+// SPDX-FileCopyrightText: 2022 Rachel Powers <508861+Ryex@users.noreply.github.com>
+//
+// SPDX-License-Identifier: GPL-3.0-only
+
+/*
+ *  Prism Launcher - Minecraft Launcher
+ *  Copyright (C) 2022 Rachel Powers <508861+Ryex@users.noreply.github.com>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "LocalWorldSaveParseTask.h"
+
+#include "FileSystem.h"
+
+#include <qdir.h>
+#include <qfileinfo.h>
+#include <quazip/quazip.h>
+#include <quazip/quazipfile.h>
+#include <quazip/quazipdir.h>
+#include <utility>
+
+namespace WorldSaveUtils {
+
+bool process(WorldSave& pack, ProcessingLevel level)
+{
+    switch (pack.type()) {
+        case ResourceType::FOLDER:
+            return WorldSaveUtils::processFolder(pack, level);
+        case ResourceType::ZIPFILE:
+            return WorldSaveUtils::processZIP(pack, level);
+        default:
+            qWarning() << "Invalid type for shader pack parse task!";
+            return false;
+    }
+}
+
+
+static std::tuple<bool, QString, bool> contains_level_dat(QDir dir, bool saves = false)
+{
+    for(auto const& entry : dir.entryInfoList()) {
+        if (!entry.isDir()) {
+            continue;
+        }
+        if (!saves && entry.fileName() == "saves") {
+            return contains_level_dat(QDir(entry.filePath()), true);
+        }
+        QFileInfo level_dat(FS::PathCombine(entry.filePath(), "level.dat"));
+        if (level_dat.exists() && level_dat.isFile()) {
+            return std::make_tuple(true, entry.fileName(), saves);
+        }
+    }
+    return std::make_tuple(false, "", saves);
+}
+
+
+bool processFolder(WorldSave& save, ProcessingLevel level)
+{
+    Q_ASSERT(save.type() == ResourceType::FOLDER);
+
+		auto [ found, save_dir_name, found_saves_dir ] = contains_level_dat(QDir(save.fileinfo().filePath()));
+
+    if (!found) {
+        return false;
+    }
+
+    save.setSaveDirName(save_dir_name);
+
+    if (found_saves_dir) {
+        save.setSaveFormat(WorldSaveFormat::MULTI);
+    } else {
+        save.setSaveFormat(WorldSaveFormat::SINGLE);
+    }
+
+    if (level == ProcessingLevel::BasicInfoOnly) {
+        return true; // only need basic info already checked
+    }
+
+    // resurved for more intensive processing
+      
+    return true; // all tests passed
+}
+
+static std::tuple<bool, QString, bool> contains_level_dat(QuaZip& zip)
```

Comment:
since you're not using a struct here, i think it would be nice if you make a comment explaining what exactly is being returned :)